CMakeLists.txt

```diff
@@ -654,6 +661,33 @@ include(cmake/OpenCVCompilerOptions.cmake)
 
 ocv_cmake_hook(POST_COMPILER_OPTIONS)
 
+# --- CUDA Support ---
+if(ENABLE_CUDA_FIRST_CLASS_LANGUAGE)
+  if(CMAKE_VERSION VERSION_LESS 3.18)
+    message(WARNING "CUDA first class language only supported for CMake versions >= 3.18, falling back to FindCUDA!")
+    set(ENABLE_CUDA_FIRST_CLASS_LANGUAGE OFF CACHE BOOL "Enable CUDA as a first class language, if enabled dependant projects will need to use CMake >= 3.18" FORCE)
+  else()
+    if(UNIX AND CUDA_PATH AND NOT $ENV{CUDA_PATH})
+      set(ENV{CUDA_PATH} ${CUDA_PATH})
+    elseif(WIN32 AND CUDA_PATH)
+      set(ENV{PATH} "${CUDA_PATH}\\bin\;$ENV{PATH}")
+    endif()
+    include(CheckLanguage)
+    check_language(CUDA)
+    # Checking windows default search location isn't possible because the CUDA Toolkit is installed to C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/vXX.X
+    if(UNIX AND NOT CMAKE_CUDA_COMPILER)
+      message(STATUS "Checking for CUDA compiler in default location (/usr/local/cuda)")
+      set(CUDA_PATH "/usr/local/cuda" CACHE INTERNAL "")
+      set(ENV{CUDA_PATH} ${CUDA_PATH})
```

Comment:
Apparently this branch doesn't work as expected. `check_language(CUDA)` doesn't detect nvcc. I've added more debug prints:
```
-- Checking for CUDA compiler in default location (/usr/local/cuda)
CMake Warning at CMakeLists.txt:685 (message):
  Detecting CUDA language (default)...  (CUDA_PATH=/usr/local/cuda / ENV{CUDA_PATH}=/usr/local/cuda)


CMake Warning at CMakeLists.txt:687 (message):
  CMAKE_CUDA_COMPILER=NOTFOUND
```
It seems that changing environment variables here do not affect `check_language(CUDA)`, I tried also prepending it's location to `PATH` also without success. While setting `CUDA_PATH` or `PATH` in the shell before running cmake helps with detection.